### DISCUSSION TITLE
Error while trying to run drive cycles long experiment on Li ion SPM model

### DISCUSSION THREAD

**Discussion by alihassan90** [pybamm-team/PyBaMM#3117] [General]
I am getting this error while i try to run the experiment on my Li-ion cell
```
step 3/5: _Step(current, [[    1     2]
 [    2     3]
 [    3     2]
 ...
 [48059     2]
 [48060     2]
 [48061     3]])
2023-07-06 16:32:27.994 - [WARNING] callbacks.on_experiment_infeasible(231): 

	Experiment is infeasible: 'event: Minimum voltage [V]' was triggered during '_Step(current, [[    1     2]
 [    2     3]
 [    3     2]
 ...
 [48059     2]
 [48060     2]
 [48061     3]])'. The returned solution only contains up to step 3 of cycle 1. 
2023-07-06 16:32:30.464 - [NOTICE] callbacks.on_experiment_end(220): Finish experiment simulation, took 3.981 s
```

The code looks like this

```python3
import pybamm
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np

parameter_values = pybamm.ParameterValues("Mohtat2020")
parameter_values.update({"SEI kinetic rate constant [m.s-1]": 1e-12})
#parameter_values.update({"SEI kinetic rate constant [m.s-1]": 1e-14})
spm = pybamm.lithium_ion.SPM({"thermal": "lumped", "SEI": "reaction limited","SEI film resistance": "distributed", "SEI porosity change": "false"})
# Calculate stoichiometries at 100% SOC

pybamm.set_logging_level("NOTICE")



drive_cycle=pd.read_csv("US06.csv",comment='#',header=None).to_numpy()

def map_drive_cycle(x, min_op_value, max_op_value):
    min_ip_value = x[:, 1].min()
    max_ip_value = x[:, 1].max()
    x[:, 1] = (x[:, 1] - min_ip_value) / (max_ip_value - min_ip_value) * (
        max_op_value - min_op_value
    ) + min_op_value
    return x

drive_cycle_power = map_drive_cycle(drive_cycle, 1.5, 3.5)



experiment = pybamm.Experiment(
    [
        (
             "Discharge at 1 C until 3.0 V",
             "Rest for 1 hour",
             pybamm.step.current(drive_cycle),
             "Charge at 1C until 4.2V",
             "Hold at 4.2V until C/50"
         ),
    ]*5,
)
```

**Comment by brosaplanella** [comment]
The Mohtat parameter set has initial conditions for a fully discharged battery. You can either add a charging step at the beginning of your experiment, or pass an `initial_soc` argument when you [solve your simulation](https://www.sciencedirect.com/science/article/pii/S0378775317307619).

**Reply by alihassan90** [reply]
Thank you, it worked.

The paper you tagged above shows how SEI and Loss of Lithium makes the degradation non-linear near the knee point or beyond 80% SOH. However I am only getting a linear degradation no matter how many cycles I run my drive cycles for. Whats the reason for that. Do i need to change the SEI, Lithium plating etc options?



![WhatsApp Image 2023-07-07 at 09 49 52](https://github.com/pybamm-team/PyBaMM/assets/38348047/a5b8e761-fbe6-40b7-ac37-7fcb6d73c983)

**Reply by brosaplanella** [reply]
Sorry, I pasted the wrong link. I meant to paste a link of the PyBaMM docs: https://docs.pybamm.org/en/latest/source/api/simulation.html#pybamm.Simulation.solve

I guess the knee is not captured due to the physics/parameters you are using. For example, in the paper above they use SEI + plating, with porosity change. With SEI only (and constant porosity) I don't think it is possible to capture the knee.

**Reply by alihassan90** [reply]
Any suggestions on how i can inculcate porosity change in my experiment?

 "SEI porosity change": "true"
or
"lithium plating porosity change": "true"

will it do the the job? Thanks

**Reply by brosaplanella** [reply]
That should enable the porosity change, but might not be enough to capture the knee. Note that if you have porosity change for plating, but don't set a plating model, that has no contribution to your model.

**Comment by ganmad** [comment]
Is there an easier way to add cut-off voltages for each step in drive profile for below command?

experiment = pybamm.Experiment( [pybamm.step.current(drive_cycle_current)] )

This command lets simulation run with a current given on one column for a duration given on another column, for a row. Now, for each step on given row, can I add a cut-off voltage as well? 

The alternative method I think is to generate a list of commands for each row, like below:

pybamm.step.current(1, duration="1 hour", termination="2.5 V")

pybamm.step.current(-1, duration="1 hour", termination="4.2 V")

But, can I use the command to run the steps between 2.5V - 4.2V, and if the step crosses the voltage, then move to next step? 
Something like  experiment = pybamm.Experiment( [pybamm.step.current(drive_cycle_current), cut-off = ["2.5 V", "4.2 V"] )

Thanks

**Reply by brosaplanella** [reply]
You can pass a termination event, or a list of termination events in your steps, see [the documentation](https://docs.pybamm.org/en/latest/source/api/experiment/experiment_steps.html) and [this example](https://docs.pybamm.org/en/latest/source/examples/notebooks/getting_started/tutorial-5-run-experiments.html) for more info.

**Reply by ganmad** [reply]
Thank you for the reply.

I have tried termination along with drive profile, as below:
experiment = pybamm.Experiment( [pybamm.step.current(drive_cycle_current), termination= ["2.5 V", "4.2 V"] )

But, this approach is stopping the experiment for the occurrence of event for first time. However, I need the experiment to continue for all the steps listed in drive profile, stopping the current step and moving to the next step. For example, the drive profile has a series of discharge, charge, rest, discharge, charge, rest steps. The approach I used stops at the first discharge step. 

am I missing something? Thanks.

**Reply by brosaplanella** [reply]
I am not sure I follow you. Does adding the termination condition to each step solve your problem?

**Reply by ganmad** [reply]
Hi Ferran,

Yes, I passed a list of commands to the Experiment class for each row in the drive profile, by creating a list using for loop:

pybamm.step.current(1, duration="1 hour", termination="2.5 V")
pybamm.step.current(-1, duration="1 hour", termination="4.2 V")
pybamm.step.current(0, duration="1 hour")
pybamm.step.current(1, duration="1 hour", termination="2.5 V")
pybamm.step.current(-1, duration="1 hour", termination="4.2 V")
pybamm.step.current(0, duration="1 hour",)
pybamm.step.current(1, duration="1 hour", termination="2.5 V")
pybamm.step.current(-1, duration="1 hour", termination="4.2 V")
...........

I will try to convey my issue clearly. For example, my drive profile has only two columns:
t   vs current
1      1
2      -1
3       0
4      1
5      -1
6      0
.       .
.       .

However, if duration of the current leads to voltage below/above cut-off limits, the simulation is being stopped there itself instead of going to next step if I pass the command : experiment = pybamm.Experiment( [pybamm.step.current(drive_cycle_current), termination= ["2.5 V", "4.2 V"] ).

So, I am wondering if there is a quick way to do this for drive profile. 

Thanks.

**Reply by brosaplanella** [reply]
So you basically want a sequence of hour-long charge/discharge steps, with a termination condition? If there is repetition, you can use the list operators `+` and `*` to simplify your notation. In the notebook I linked above there are examples.

**Reply by ganmad** [reply]
Agreed. Thanks for taking time to reply. Sorry, I over simplified current profile, but I was asking what could we do if duration and magnitude of current varies frequently as in drive cycle. If the cell can hold the current for the given duration, there won't be a problem running the drive profile on the cell. Otherwise, it wont run for the given duration.

**Reply by brosaplanella** [reply]
If you run a given drive cycle (e.g. in which current changes every second), then you can use 
```python3
experiment = pybamm.Experiment([pybamm.step.current(drive_cycle)])
```
where `drive_cycle` is a 2-column array (more details in [Tutorial 5](https://docs.pybamm.org/en/latest/source/examples/notebooks/getting_started/tutorial-5-run-experiments.html)). If you set termination conditions they will apply to the whole drive cycle. 

If you want the drive cycle to continue from the next step, then I think you need to define each individual step as you did above.

**Reply by alihassan90** [reply]
Hello, Is there a way to include both termination conditions

max,min voltage
&
80% SOH in experiment. Apparently "termination" key word can be used once only.

**Reply by alihassan90** [reply]
Also this pybamm.Experiment( [pybamm.step.current(drive_cycle_current), termination= ["2.5 V", "4.2 V"] is treating 4.2V as stopping voltage and not maximum voltage. web

Error:

2023-07-26 11:33:06.084 - [NOTICE] callbacks.on_cycle_end(213): Stopping experiment since minimum voltage (3.933 V) is below stopping voltage (4.200 V).
2023-07-26 11:33:06.086 - [NOTICE] callbacks.on_experiment_end(220): Finish experiment simulation, took 25.471 s

**Reply by ganmad** [reply]
Instead of drive profile, you can create a list of step current commands using for loop for each row in the drive profile and pass them to Experiment class. You can then use termination as 80% capacity as well. Hope this works.

For example: For each row in the drive profile,
pybamm.step.current(1, duration="1 hour", termination="2.5 V")
pybamm.step.current(-1, duration="1 hour", termination="4.2 V")
pybamm.step.current(0, duration="1 hour")
pybamm.step.current(1, duration="1 hour", termination="2.5 V")
pybamm.step.current(-1, duration="1 hour", termination="4.2 V")
pybamm.step.current(0, duration="1 hour",)
pybamm.step.current(1, duration="1 hour", termination="2.5 V")
pybamm.step.current(-1, duration="1 hour", termination="4.2 V")

**Comment by ganmad** [comment]
Is there a way to add SoC as a cut-off instead of voltage for discharge/charge step? 
For example,  I wanted to implement:
pybamm.step.current(1, duration="1 hour", termination="5% Capacity")

I wanted to run several of such charge/discharge cycles. I understand termination works for whole cycling experiment, but not sure how to cycle the cell between partial depth of discharge, say 80% to 20% SoC instead of voltage cut-off.  Any insights will be helpful.  Thanks.

Best regards,
Ganesh

**Reply by brosaplanella** [reply]
This is not yet implemented, but we are working on it: https://github.com/pybamm-team/PyBaMM/issues/2553

**Reply by ganmad** [reply]
Thanks for update.
